### PR TITLE
bsunanda:Run2-alca51 Add AlCaReco and DQM code for HcalIsolatedBunch

### PR DESCRIPTION
**Pull request by bsunanda**
A selector function is added in AlCaReco to select HLT's from HcalIsolatedBunch and also the needed cfi/cf files. Some DQM code is added to monitor these

### CODE REVIEW

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by mmusich** [comment]
@bsunanda few comments questions.
If this has to be run at Tier-0 you will need to update also:

Configuration/StandardSequences/python/AlCaRecoStreams_cff.py
Configuration/EventContent/python/AlCaRecoOutput_cff.py
Configuration/AlCa/python/autoAlCa.py

just as you did for HcalCalIsoTrkFilter in #13251.

Also, what is the expected data volume generated by this AlCaReco?

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@mmusich @dmitrijus Could you check this and approve this?

**Comment by dmitrijus** [comment]
+1

**Comment by mmusich** [comment]
+1

File: Calibration/HcalAlCaRecoProducers/plugins/AlCaIsolatedBunchSelector.cc [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+  if (!triggerEventHandle.isValid()) {
+    edm::LogWarning("HcalIsoBunch") << "Error! Can't get the product "
+				    << triggerEvent_.label() ;
+  } else {
+    triggerEvent = *(triggerEventHandle.product());
+    /////////////////////////////TriggerResults
+    edm::Handle<edm::TriggerResults> triggerResults;
+    iEvent.getByToken(tok_trigRes_, triggerResults);
+    if (triggerResults.isValid()) {
+      std::vector<std::string> modules;
```
Comment (davidlange6) [inline Calibration/HcalAlCaRecoProducers/plugins/AlCaIsolatedBunchSelector.cc]: it seems this variable is never used (and similarly below in the code where it is repeated)
@bsunanda

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@mmusich @dmitrijus We do need your approval again

**Comment by mmusich** [comment]
Hi @bsunanda let's see the output of the tests

File: Calibration/HcalAlCaRecoProducers/plugins/AlCaIsolatedBunchFilter.cc [hunk 2]
```diff
@@ -115,63 +110,54 @@ bool AlCaIsolatedBunchFilter::filter(edm::Event& iEvent,
...
-	    for (unsigned int i=0; i<trigJetNames_.size(); ++i) {
-	      if (triggerNames_[iHLT].find(trigJetNames_[i].c_str()) !=
-		  std::string::npos) {
-		if (hlt > 0) jet = true;	
-		if (jet) {
+    /////////////////////////////TriggerResults
+    edm::Handle<edm::TriggerResults> triggerResults;
+    iEvent.getByToken(tok_trigRes_, triggerResults);
+    if (triggerResults.isValid()) {
+      std::vector<std::string> modules;
```
Comment (mmusich) [inline Calibration/HcalAlCaRecoProducers/plugins/AlCaIsolatedBunchFilter.cc]: @bsunanda still looks like this vector of strings `std::vector<std::string> modules` is not used anywhere

File: Calibration/HcalAlCaRecoProducers/plugins/AlCaIsolatedBunchSelector.cc [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+  edm::LogInfo("AlCaIsoBunch") << "Run " << iEvent.id().run() << " Event " 
+			       << iEvent.id().event() << " Luminosity " 
+			       << iEvent.luminosityBlock() << " Bunch " 
+			       << iEvent.bunchCrossing() << std::endl;
+#endif
+  //Step1: Find if the event passes the chosen trigger
+  edm::Handle<edm::TriggerResults> triggerResults;
+  iEvent.getByToken(tok_trigRes_, triggerResults);
+  if (triggerResults.isValid()) {
+    std::vector<std::string> modules;
```
Comment (mmusich) [inline Calibration/HcalAlCaRecoProducers/plugins/AlCaIsolatedBunchSelector.cc]: here as well

File: DQMOffline/CalibCalo/src/DQMHcalIsolatedBunchAlCaReco.cc [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+void DQMHcalIsolatedBunchAlCaReco::analyze(const edm::Event& iEvent, 
+					   const edm::EventSetup& iSetup ){  
+ 
+  bool accept(false);
+  /////////////////////////////TriggerResults
+  edm::Handle<edm::TriggerResults> triggerResults;
+  iEvent.getByToken(trigResult_, triggerResults);
+  if (triggerResults.isValid()) {
+    std::vector<std::string> modules;
```
Comment (mmusich) [inline DQMOffline/CalibCalo/src/DQMHcalIsolatedBunchAlCaReco.cc]: and here too

**Comment by mmusich** [comment]
please test

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by dmitrijus** [comment]
+1

**Comment by mmusich** [comment]
+1

**Comment by bsunanda** [comment]
@davidlange6 Could you kindly merge this? We most likely would get the isolated bunch triggers this weekend and need this and back ported to 8_0_0 soon

File: DQMOffline/CalibCalo/src/DQMHcalIsolatedBunchAlCaReco.cc [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+  bool accept(false);
+  /////////////////////////////TriggerResults
+  edm::Handle<edm::TriggerResults> triggerResults;
+  iEvent.getByToken(trigResult_, triggerResults);
+  if (triggerResults.isValid()) {
+    const edm::TriggerNames & triggerNames = iEvent.triggerNames(*triggerResults);
+    const std::vector<std::string> & triggerNames_ = triggerNames.triggerNames();
+    for (unsigned int iHLT=0; iHLT<triggerResults->size(); iHLT++) {
+      int hlt    = triggerResults->accept(iHLT);
+      if (triggerNames_[iHLT].find(trigName_.c_str())!=std::string::npos) {
```
Comment (davidlange6) [inline DQMOffline/CalibCalo/src/DQMHcalIsolatedBunchAlCaReco.cc]: @bsunanda - do this string comparison only if hlt>0

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by davidlange6** [comment]
whoops I merged the 80x by accident - should be ok -but lets see...

**Comment by mmusich** [comment]
+1

**Comment by bsunanda** [comment]
@dmitrijus requesting once again for the revised version

**Pull request merged**